pkg/ingester/metrics.go

```diff
@@ -0,0 +1,187 @@
+			Name: "cortex_ingester_queries_total",
+			Help: "The total number of queries the ingester has handled.",
+		}),
+		queriedSamples: prometheus.NewHistogram(prometheus.HistogramOpts{
+			Name: "cortex_ingester_queried_samples",
+			Help: "The total number of samples returned from queries.",
+			// Could easily return 10m samples per query - 10*(8^(8-1)) = 20.9m.
+			Buckets: prometheus.ExponentialBuckets(10, 8, 8),
+		}),
+		queriedSeries: prometheus.NewHistogram(prometheus.HistogramOpts{
+			Name: "cortex_ingester_queried_series",
+			Help: "The total number of series returned from queries.",
+			// A reasonable upper bound is around 100k - 10*(8^(6-1)) = 327k.
+			Buckets: prometheus.ExponentialBuckets(10, 8, 6),
+		}),
+		queriedChunks: prometheus.NewHistogram(prometheus.HistogramOpts{
+			Name: "cortex_ingester_queried_chunks",
+			Help: "The total number of chunks returned from queries.",
+			// A small number of chunks per series - 10*(8^(7-1)) = 2.6m.
+			Buckets: prometheus.ExponentialBuckets(10, 8, 7),
+		}),
+	}
+
+	if r != nil {
+		r.MustRegister(
+			m.flushQueueLength,
+			m.ingestedSamples,
+			m.ingestedSamplesFail,
+			m.queries,
+			m.queriedSamples,
+			m.queriedSeries,
+			m.queriedChunks,
+		)
+	}
+
+	return m
+}
+
+// TSDB shipper metrics. We aggregate metrics from individual TSDB shippers into
+// a single set of counters, which are exposed as Cortex metrics.
+type shipperMetrics struct {
+	dirSyncs        *prometheus.Desc // sum(thanos_shipper_dir_syncs_total)
+	dirSyncFailures *prometheus.Desc // sum(thanos_shipper_dir_sync_failures_total)
+	uploads         *prometheus.Desc // sum(thanos_shipper_uploads_total)
+	uploadFailures  *prometheus.Desc // sum(thanos_shipper_upload_failures_total)
+
+	regsMu sync.RWMutex                    // custom mutex for shipper registry, to avoid blocking main user state mutex on collection
+	regs   map[string]*prometheus.Registry // One prometheus registry (used by shipper) per tenant
+}
+
+func newShipperMetrics(r prometheus.Registerer) *shipperMetrics {
+	m := &shipperMetrics{
+		regs: make(map[string]*prometheus.Registry),
+
+		dirSyncs: prometheus.NewDesc(
+			"cortex_ingester_shipper_dir_syncs_total",
+			"TSDB: Total dir sync attempts",
+			nil, nil),
+		dirSyncFailures: prometheus.NewDesc(
+			"cortex_ingester_shipper_dir_sync_failures_total",
+			"TSDB: Total number of failed dir syncs",
+			nil, nil),
+		uploads: prometheus.NewDesc(
+			"cortex_ingester_shipper_uploads_total",
+			"TSDB: Total object upload attempts",
+			nil, nil),
+		uploadFailures: prometheus.NewDesc(
+			"cortex_ingester_shipper_upload_failures_total",
+			"TSDB: Total number of failed object uploads",
+			nil, nil),
+	}
+
+	if r != nil {
+		r.MustRegister(m)
+	}
+	return m
+}
+
+func (sm *shipperMetrics) Describe(out chan<- *prometheus.Desc) {
+	out <- sm.dirSyncs
+	out <- sm.dirSyncFailures
+	out <- sm.uploads
+	out <- sm.uploadFailures
+}
+
+func (sm *shipperMetrics) Collect(out chan<- prometheus.Metric) {
```

Comment:
Have you considered adding a test on `shipperMetrics.Collect()` to cover at least the happy path? It shouldn't be much complicated if we use `testutil.GatherAndCompare()` to test on it.

Reply:
Added test now.